Fix warnings in the jitter code

janvorli opened the pull request:
This change fixes some of the warnings in the jitter code on Linux. With this change
combined with an upcoming change in the rest of the codebase, 16 warning disabling
options can be removed.

The issues in the jitter were:
1) Incorrect typedefs for const_reference and const_pointer (the const was ignored)
2) Member initiazalization order in class constructor didn't correspond to the member order in some classes
3) Objects with vtables were copied via memcpy. While this is intentional, cast of the pointers to void*
   is required to silence clang warning
4) Comparing values of different enums - there are cases of two enums containing the same values.
5) Casting int to pointer - the cast was legal (forming a handle), adding cast to size_t in between used to fix the warning
6) "static struct" - removed the static keyword
7) Missing return value in methods with _ASSERTE
8) Class name classifier on methods in the .h
9) Specialized template member functions need to be defined out of the class

janvorli commented:
@cmckinsey, @CarolEidt  Can you take a look please?

cmckinsey commented on src/jit/codegencommon.cpp:
```diff
@@ -87,9 +87,9 @@ CodeGenInterface *getCodeGenerator(Compiler *comp)
     compiler(theCompiler),
-    gcInfo(theCompiler),
-    regSet(theCompiler, gcInfo)
+    gcInfo(theCompiler)
```
I think we should instead reorder the definitions in the class so that gcInfo is before regSet. (Notice that regSet takes the gcInfo as an argument). Not sure if there is a real problem but seems like regSet should be able to assume it has an initialized gcInfo.

CarolEidt replied:
I agree with Chris - it would be cleaner, even if the order is not critical.

cmckinsey commented:
The rest is LGTM.

BruceForstall commented on src/jit/compiler.hpp:
```diff
@@ -1341,7 +1341,8 @@ void                GenTree::CopyFrom(const GenTree* src, Compiler* comp)
     GenTreePtr prev = gtPrev;
     GenTreePtr next = gtNext;
-    memcpy(this, src, src->GetNodeSize());
+    // The VTable is copied intentionally here
```
technically, the vtable pointer

BruceForstall commented on src/jit/lsra.h:
```diff
@@ -503,7 +503,7 @@ class LinearScan : public LinearScanInterface
                                       LSRA_RELOAD_MASK                  = 0x400 };
     LsraReload                  getLsraReload()                 { return (LsraReload) (lsraStressMask & LSRA_RELOAD_MASK); }
-    bool                        alwaysInsertReload()            { return getLsraSpill() == LSRA_ALWAYS_INSERT_RELOAD; }
+    bool                        alwaysInsertReload()            { return getLsraReload() == LSRA_ALWAYS_INSERT_RELOAD; }
```
fyi, Carol has already fixed this. maybe not checked in yet?

CarolEidt replied:
It exposes another bug (only exposed by stress mode), for which I'm testing a fix now.  Meanwhile it's fine to check this in.

BruceForstall commented:
LGTM

janvorli commented:
Fixed the order of gcInfo and regSet as per @cmckinsey's comment, fixed the comment in the compiler.hpp as per @BruceForstall's comment. Squashed.
I will merge it in once the tests complete.
